client/mysqltest.cc

```diff
@@ -3837,44 +3837,337 @@ void do_remove_files_wildcard(struct st_command *command)
 
   DESCRIPTION
-  copy_file <from_file> <to_file>
-  Copy <from_file> to <to_file>
+  copy_file <source> <destination> [-r]
 
-  NOTE! Will fail if <to_file> exists
+  <source> can be a file, a directory, or a path with wildcard pattern
+
+  - When <source> is a directory, '-r' must be given to copy recursively
+  - When <source> is a wildcard pattern, '-r' is optional. If given, then
+    directories that match the pattern will also be copied.
 */
 
 void do_copy_file(struct st_command *command)
 {
   int error;
-  static DYNAMIC_STRING ds_from_file;
-  static DYNAMIC_STRING ds_to_file;
-  const struct command_arg copy_file_args[] = {
-    { "from_file", ARG_STRING, TRUE, &ds_from_file, "Filename to copy from" },
-    { "to_file", ARG_STRING, TRUE, &ds_to_file, "Filename to copy to" }
+  static DYNAMIC_STRING ds_source;
+  static DYNAMIC_STRING ds_destination;
+  static DYNAMIC_STRING ds_wild;
+  static DYNAMIC_STRING ds_recursive;
+  static bool recursive;
+  MY_STAT stat_buff;
+  const struct command_arg copy_file_args[]=
+  {
+    { "source", ARG_STRING, TRUE, &ds_source, "File/dir name to copy from" },
+    { "destination", ARG_STRING, TRUE, &ds_destination,
+      "File/dir name to copy to" },
+    { "recursive", ARG_STRING, FALSE, &ds_recursive,
+      "Optional argument indicating if directories should be copied recursively" }
   };
   DBUG_ENTER("do_copy_file");
 
   check_command_args(command, command->first_argument,
                      copy_file_args,
                      sizeof(copy_file_args)/sizeof(struct command_arg),
                      ' ');
+  recursive= !strcmp("-r", ds_recursive.str);
 
-  if (bad_path(ds_to_file.str))
-    DBUG_VOID_RETURN;
+  /*
+    Throw an error if destination path is invalid, or source and destination
+    paths are same.
+  */
+  if (bad_path(ds_destination.str) ||
+      !strcmp(ds_source.str, ds_destination.str))
+  {
+    error= 1;
+  }
+  else if (my_stat(ds_source.str, &stat_buff, MYF(0)))
+  {
+    /* For regular file, copy directly */
+    if (MY_S_ISREG(stat_buff.st_mode))
+    {
+      DBUG_PRINT("info", ("Copy %s to %s", ds_source.str, ds_destination.str));
+      /* MY_HOLD_ORIGINAL_MODES prevents attempts to chown the file */
+      error= (my_copy(ds_source.str, ds_destination.str,
+              MYF(MY_DONT_OVERWRITE_FILE | MY_WME | MY_HOLD_ORIGINAL_MODES)) != 0);
+    }
+    /* For directory, copy if recursive is enabled */
+    else if (recursive)
+    {
+      DBUG_PRINT("info", ("Recursive copy files of %s to %s",
+                          ds_source.str, ds_destination.str));
+      DBUG_PRINT("info", ("listing directory: %s", ds_source.str));
+      error= recursive_copy(&ds_source, &ds_destination);
+    }
+    else
+    {
+      error= 1;
+    }
+  }
+  else
+  {
+    /*
+      Cannot get file stat. Could be a wildcard pattern. Extract the basename
+      as wildcard pattern.  If it is not a vaild wildcard pattern,
+      wildcard_copy() will return error.
```

Comment:
so, `copy_file somewhere/*/* dst` will not work.

This is fine, but better to mention it somewhere, e.g in a comment before the function. That only the last component of the src path can be a wildcard.